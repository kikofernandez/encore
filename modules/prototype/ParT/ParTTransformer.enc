module ParTTransformer (
    arrT,
    firstT,
    composeT)

import ParT.ParT
import ParT.Arrow as A

typedef Transformer[a, b] = PArrow[a, Par[(Par[b], Par [a -> Par[b]])]]

fun arrT[a, b](fn: PArrow[a, Par[b]]): Transformer[a, b]
  val transformer = fun (x: a) => join(runArrow(x, fn))
  -- issue: `A.arr` is synonym for `liftv` but it does not typecheck
  A.arr(fun (y: a) => liftv((transformer(y), A.arr(transformer))))
end

fun firstT[a, b, c](fn: PArrow[a, Par[b]]): Transformer[(a, c), (b, c)]
  val transformer = fun (t: (a, c)) => join(runArrow(t.0, fn)) >> fun (x: b) => (x, t.1)
  A.arr(fun (t: (a, c)) => liftv((transformer(t), A.arr(transformer))))
end

-- Transformer[a, b] = PArrow[a, Par[(Par[b], Par [a -> Par[b]])]]
-- Transformer[b, c] = PArrow[b, Par[(Par[c], Par [b -> Par[c]])]]
-- Transformer[a, c] = PArrow[a, Par[(Par[c], Par [a -> Par[c]])]]
fun composeT[a, b, c](t1: Transformer[a, b], t2: Transformer[b, c]): Transformer[a, c]
  val transformer = fun (x: a)
                      val pTuple = join(runArrow(x, t1)) -- Par[(Par[b], Par [a -> Par[b]])]
                      val clos = fun (tuple: (Par[b], PArrow[a, Par[b]]))
                                   val pb = tuple.0
                                   val inter = join(runParArrow(pb, t2)) -- Par[(Par[c], Par [b -> Par[c]])]
                                   inter >> (fun (tInterm: (Par[c], Par[b -> Par[c]]))
                                              (tInterm.0, A.composePar(tuple.1, tInterm.1))
                                             end)
                                 end
                      bind(clos, pTuple) -- (a -> Par[b]) ->  Par[a] -> Par[b]
                    end
  liftv(transformer)
end

fun runTrans[a, b](item: a, t: Transformer[a, b]): Par[(Par[b], Transformer[a, b])]
  val pp = join(runArrow(item, t))  -- Par[(Par[b], Par[a -> Par[b]])]
  pp >> (fun (p: (Par[b], Par[a -> Par[b]]))
          (p.0, arrT(p.1))
         end)
end

fun runTransT[a, b](item: a, t: Transformer[a, b]): Par[b]
  val pp = join(runArrow(item, t))
  bind(fun (p: (Par[b], Par[a -> Par[b]])) => p.0, pp)
end

fun runParTrans[a, b](p: Par[a], t: Transformer[a, b]): Par[b]
  bind(fun (x: a) => runTransT(x, t), p)
end

fun fixpoint[a](t: Transformer[a, a], p: Par[a], fn: a -> bool): Par[a]
  while true do
    print(".")
    val result = runParTrans(p, t)
    -- TODO: better if this was not a synchronisation point
    val b = get(all(result, fn))
    if b then
      return result
    end
  end
  return empty[a]()
end

fun inc(x: int): Par[int]
  liftv(x + 1)
end

fun double(x: int): Par[int]
  liftv(x * x)
end

fun convertToTrue(x: bool): Par[bool]
  if not x then
    liftv(true)
  else
    liftv(x)
  end
end

fun flipBool(x: bool): Par[bool]
  liftv(not x)
end

-- TODO: add the switch combinator (some paper, for runtime switch of streams)
-- TODO: add the kill combinator which kills streams running ParTs


class Main
  def main(): unit
    -- val arrowInc = A.arr(inc)
    -- val arrowDouble = A.arr(double)
    -- val transformerInc = arrT(arrowInc)
    -- val transformerDouble = arrT(arrowDouble)
    -- val transformer = composeT(transformerInc, transformerDouble)
    -- val p = runTrans(3, transformer)
    -- p >> (fun (t: (Par[int], Transformer[int, int]))
    --         (t.0) >> fun (x: int) => runTrans(x, t.1)
    --       end)

    println("Example.enc")
    val pp = liftv(false) ||| liftv(true)
    val convertB = A.arr(convertToTrue)
    val transformerB = arrT(convertB)
    fixpoint(transformerB, pp, fun (b: bool) => b)

    val flip = A.arr(flipBool)
    val transformerFlip = arrT(flip)
    fixpoint(transformerFlip, pp, fun (b: bool) => b)
  end
end
