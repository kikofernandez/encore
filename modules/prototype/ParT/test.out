Test: sequence.
Number: 1
Number: 0
Test: distinct.
Number: 34
Number: 12
Test: each.
Number: 1
Number: 2
Number: 3
Number: 4
Number: 5
Number: 6
Number: 7
Number: 8
Number: 9
Number: 10
Number: 13
Number: 1
Number: 2
Number: 3
Number: 4
Number: 5
Number: 6
Number: 7
Number: 8
Number: 9
Number: 10
Number: 13
Test: each distinct
Number: 35
Number: 13
Test: recomputation.
Start computation...
Start computation...
Number: 13
Number: 13
Test: cache.
Start computation...
Number: 14
Number: 14
