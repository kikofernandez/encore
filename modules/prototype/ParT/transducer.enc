fun comp[t, t', t''](f: t' -> t'', g: t -> t'): t -> t''
  fun (x: t) => f(g(x))
end

typedef Trans = ((int, int) -> int) -> ((int, int) -> int)

fun map(fn : int -> int): Trans
  fun (rf : (int, int) -> int): (int, int) -> int
    fun (result: int, input: int) => rf(result, fn(input))
  end
end

fun take(items: int): ((int, int) -> int) -> ((int, int) -> int)
  fun (rf : (int, int) -> int) : (int, int) -> int
    fun (result: int, input: int) : int
      EMBED (unit)
      _enc__field_items = __atomic_fetch_sub(&((struct _enc__env_closure3*) _env)->_enc__field_items, 1, __ATOMIC_ACQ_REL);
      END
      if (items >= 1) then
        rf(result, input)
      else
        result
      end
    end
  end
end

-- not that easy because of the types...
-- fun bind(fn: int -> [int]): ((int, int) -> int) -> ((int, int) -> int)
-- end

fun filtr(pred : int -> bool): Trans
  fun (rf : (int, int) -> int): (int, int) -> int
    fun (result: int, input: int) : int
      if pred(input) then
        rf(result, input)
      else
        result
      end
    end
  end
end

-- fun delay(rf: (int, int) -> int): ((int, int) -> int)
--     fun (result: int, input: int): int
--       rf(result, input)
--     end
--   end
-- end

-- fun resume(fn: () -> ((int, int) -> int)): Trans
--   fun (rf: (int, int) -> int): ((int, int) -> int)
--     fun (result: int, input: int): int
--        val extractedFn = fn()
--        extractedFn(result, input)
--     end
--   end
-- end

fun inc(x: int): int
  x + 5
end

fun dec(x: int): int
  x -1
end

fun sum(acc: int, input: int): int
  acc + input
end

fun reduc[t, t'](coll: [t], f: (t', t) -> t', var init: t'): t'
  for item <- coll do
    init = f(init, item)
  end
  return init
end

fun step(xform: Trans, f: (int, int) -> int, init: int, coll: [int]): (int, () -> int)
  val f = xform(f)
  val index = 0
  val size = |coll|
  val result = f(init, coll(index))
  val fn = fun (): int
             EMBED (unit)
               _enc__field_index = __atomic_fetch_add(&((struct _enc__env_closure0*) _env)->_enc__field_index, 1, __ATOMIC_RELAXED);
               _enc__field_result = __atomic_load_n(&((struct _enc__env_closure0*) _env)->_enc__field_result, __ATOMIC_RELAXED);
             END
             val result = f(result, coll(index))
             EMBED (unit)
               ((struct _enc__env_closure0*) _env)->_enc__field_result = #{result};
             END
             result
           end
  (result, fn)
end

fun odd(x: int): bool
  if x % 2 == 0 then
    false
  else
    true
  end
end

fun transduce(xform: Trans, f: (int, int) -> int, init: int, coll: [int]): int
  val f = xform(f)
  reduc(coll, f, init)
end

active class Main
  def main(): unit
    val xform = comp(map(inc), map(dec))
    -- val xxform = comp(filtr(odd), xform)
    val xxform = comp(take(2), xform)
    -- val xform2 = comp(resume, xform1)
    -- val fn = xform(sum)
    -- println(fn())
    -- val xxform = comp(resume, xform)

    -- Functional transducer, can be shared!
    -- In Clojure, they keep state around, which means, do not share
    -- your xform, use and discard.
    println(transduce(xxform, sum, 0, [1, 2, 3]))

    val oneT = step(xform, sum, 0, [1, 2, 3])
    val fn = oneT.1
    val two = fn()
    println("First item: {}\t second: {}", oneT.0, two)
  end
end
