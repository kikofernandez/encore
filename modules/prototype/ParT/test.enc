import ParT.ParT

fun gen(): Par[int]
  liftv(34)
end

fun gen2(): Par[int]
  liftv(12)
end

fun gen3(): Par[int]
  println("Start computation...")
  liftv(12)
end

fun eachBasedParT(): Par[int]
  val ar = new [int](10)
  for i <- [0..9] do
    ar(i) = i
  end
  each(ar)
end

fun inc(x: int): int
  x + 1
end

fun adult(age: int): bool
  if age < 18 then
    false
  else
    true
  end
end

fun cmp(x: int, y: int): int
  if x < y then
    -1
  else if x > y then
    1
  else
    0
  end
end

active class Main
  def distinct(): unit
    println("Test: distinct.")

    val p11 = gen
    val p12 = gen2
    val p3 = EMBED ([int])
               delayed_par_t *ast = new_delay_par_value(_ctx, (delay_t*) #{p11}, ENCORE_PRIMITIVE);
               delayed_par_t *ast2 = new_delay_par_value(_ctx, (delay_t*) #{p12}, ENCORE_PRIMITIVE);
               delayed_par_t *astMerged = new_delayed_par_merge(_ctx, ast, ast2, &party_type);

               delayed_par_t *ast3 = new_delay_par_value(_ctx, (delay_t*) #{p12}, ENCORE_PRIMITIVE);
               delayed_par_t *astMerged2 = new_delayed_par_merge(_ctx, astMerged, ast3, &party_type);

               ast = delay_distinct(_ctx, astMerged2, #{cmp}, ENCORE_PRIMITIVE);
               par_t* p = delay_extract(_ctx, ast, ENCORE_PRIMITIVE);
               p;
             END

    for x <- p3 do
      println("Number: {}", x)
    end
  end

  def each(): unit
    println("Test: each.")

    val p11 = eachBasedParT
    val p12 = gen2
    val p3 = EMBED ([int])
               delayed_par_t *ast = new_delay_par_value(_ctx, (delay_t*) #{p11}, ENCORE_PRIMITIVE);
               delayed_par_t *ast2 = new_delay_par_value(_ctx, (delay_t*) #{p12}, ENCORE_PRIMITIVE);
               delayed_par_t *ast3 = new_delayed_par_merge(_ctx, ast, ast2, &party_type);
               delayed_par_t *ast3Duplicated = new_delayed_par_merge(_ctx, ast3, ast3, &party_type);
               ast = delay_sequence(_ctx, ast3Duplicated, #{inc}, ENCORE_PRIMITIVE);
               par_t* p = delay_extract(_ctx, ast, ENCORE_PRIMITIVE);
               p;
             END

    for x <- p3 do
      println("Number: {}", x)
    end
  end

  def eachDistinct(): unit
    -- Testing each and distinct
    println("Test: each distinct")
    val p11 = gen
    val p12 = gen2
    val p3 = EMBED ([int])
               delayed_par_t *ast = new_delay_par_value(_ctx, (delay_t*) #{p11}, ENCORE_PRIMITIVE);
               delayed_par_t *ast2 = new_delay_par_value(_ctx, (delay_t*) #{p12}, ENCORE_PRIMITIVE);
               delayed_par_t *ast3 = new_delayed_par_merge(_ctx, ast, ast2, &party_type);
               delayed_par_t *ast3Duplicated = new_delayed_par_merge(_ctx, ast3, ast3, &party_type);
               ast = delay_sequence(_ctx, ast3Duplicated, #{inc}, ENCORE_PRIMITIVE);

               ast = delay_distinct(_ctx, ast, #{cmp}, ENCORE_PRIMITIVE);
               par_t* p = delay_extract(_ctx, ast, ENCORE_PRIMITIVE);
               p;
             END
    for x <- p3 do
      println("Number: {}", x)
    end
  end

  def sequence(): unit
    println("Test: sequence.")
    val p11 = gen
    val p12 = gen2
    val p3 = EMBED ([int])
               delayed_par_t *ast = new_delay_par_value(_ctx, (delay_t*) #{p11}, ENCORE_PRIMITIVE);
               delayed_par_t *ast2 = new_delay_par_value(_ctx, (delay_t*) #{p12}, ENCORE_PRIMITIVE);
               delayed_par_t *ast3 = new_delayed_par_merge(_ctx, ast, ast2, &party_type);
               ast = delay_sequence(_ctx, ast3, #{inc}, ENCORE_PRIMITIVE);
               ast = delay_sequence(_ctx, ast3, #{adult}, ENCORE_PRIMITIVE);
               par_t* p = delay_extract(_ctx, ast, ENCORE_PRIMITIVE);
               p;
             END
    for x <- p3 do
      println("Number: {}", x)
    end
  end

  def testRecomputation(): unit
    println("Test: recomputation.")
    val p11 = gen3
    val p3 = EMBED ([int])
               delayed_par_t *ast = new_delay_par_value(_ctx, (delay_t*) #{p11}, ENCORE_PRIMITIVE);
               delayed_par_t *ast3 = new_delayed_par_merge(_ctx, ast, ast, &party_type);
               delayed_par_t *result = delay_sequence(_ctx, ast3, #{inc}, ENCORE_PRIMITIVE);
               par_t* p = delay_extract(_ctx, result, ENCORE_PRIMITIVE);
               p;
             END
    for x <- p3 do
      println("Number: {}", x)
    end
  end

  def testCache(): unit
    println("Test: cache.")
    val p11 = gen3
    val p3 = EMBED ([int])
               delayed_par_t *ast = delay_cache(_ctx, new_delay_par_value(_ctx, (delay_t*) #{p11}, ENCORE_PRIMITIVE));
               delayed_par_t *ast3 = new_delayed_par_merge(_ctx, ast, ast, &party_type);
               delayed_par_t *result = delay_sequence(_ctx, ast3, #{inc}, ENCORE_PRIMITIVE);
               result = delay_sequence(_ctx, result, #{inc}, ENCORE_PRIMITIVE);
               par_t* p = delay_extract(_ctx, result, ENCORE_PRIMITIVE);
               p;
             END
    for x <- p3 do
      println("Number: {}", x)
    end
  end

  def main(): unit
    this.sequence()
    this.distinct()
    this.each()
    this.eachDistinct()
    this.testRecomputation()
    this.testCache()
  end
end


-- active class Main
--   def main(): unit
--     liftv(3) ||| liftv(2)
--   end
-- end
