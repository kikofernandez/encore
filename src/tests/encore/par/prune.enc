import ParT.ParT
import Task

read class MyObject
  val id : int = 0

  def init(id: int): unit
    this.id = id
  end

  def value(): int
    this.id
  end

  def inc(): uint
    EMBED (uint)
      // safely cheating to increment the counter
      #{this.id} = #{this.id} + 1;
    END
  end
end

--
-- Helper function
--
fun generateObjects(number: uint): Par[MyObject]
  var p = empty[MyObject]()
  for i <- [0..number] do
    p = p ||| liftf(async(new MyObject(i)))
  end
  p
end

--
-- TESTS
--

fun testOnlyOneItemGetsThrough(): unit
  println("\n- Tests only one item gets through:\n")
  val p = generateObjects(1000) >> fun (m : MyObject) => m.id
  prune(fun (f : Fut[Maybe[int]])
          join(liftf(f ~~> fun (m : Maybe[int])
                             match m with
                               case Just(i) =>
                                 do
                                   println("\t* Streaming counter with id: {}", i)
                                   liftv(i)
                                 end
                               end
                             end
                           end))
        end, p)
end

fun testLoopingAndPrune(): unit
  println("\n- Tests looping and pruning (not cancelling operations):\n")
  val p = generateObjects(1000)
  val p' = loop(p, fun (c: MyObject)
                     c.inc()
                     liftv(c)
                   end,
                   fun (c: MyObject) => c.id >= 500)
  extract(p')
  prune(fun (f : Fut[Maybe[MyObject]])
          join(liftf(f ~~> fun (m : Maybe[MyObject])
                             match m with
                               case Just(c) =>
                                 do
                                   println("\t* Streaming counter with id: {}", c.id)
                                   liftv(c.id)
                                 end
                               end
                             end
                           end))
        end, p')
end

active class Main
  def main(): unit
    testOnlyOneItemGetsThrough()

    testLoopingAndPrune()
  end
end
