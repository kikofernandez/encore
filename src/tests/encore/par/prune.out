
- Tests only one item gets through:

	* Streaming counter with id: 0

- Tests looping and pruning (not cancelling operations):

	* Streaming counter with id: 500
